lib/semian/circuit_breaker.rb

```diff
@@ -0,0 +1,131 @@
+module Semian
+  class CircuitBreaker
+    extend Forwardable
+    attr_reader :state
+
+    def_delegators :@resource, :acquire, :destroy, :count, :semid
+
+    def initialize(resource, exceptions:, success_threshold:, error_threshold:, error_timeout:)
+      @resource = resource
+      @success_count_threshold = success_threshold
+      @error_count_threshold = error_threshold
+      @error_timeout = error_timeout
+      @exceptions = Array(exceptions) + [::Semian::BaseError]
+      reset
+    end
+
+    def with_fallback(fallback, &block)
```

Comment:
I know we talked about it in person, but now that I see it, I think the class design was cleaner the way it was before (in the Shopify codebase) with the Resiliency::Runner. A CircuitBreaker is a concept that in itself doesn't (shouldn't) know anything about "running code" or "fallbacks" or "exceptions". This class should encapsulate the internal state of the CircuitBreaker and nothing else, imho.